python/xoscar/collective/tests/test_cupy.py

```diff
@@ -0,0 +1,313 @@
+
+
+def worker_barrier(rank, nccl_id, device_id):
+    cupy.cuda.Device(device_id).use()
+
+    context = xc.Context(n_devices=2, nccl_unique_id=nccl_id, rank=rank)
+    sendbuf = cupy.array([[1, 2, 3], [1, 2, 3]], dtype=cupy.float32)
+    recvbuf = cupy.zeros_like(sendbuf, dtype=cupy.float32)
+    op = xc.ReduceOp.SUM
+
+    xc.allreduce(context, sendbuf, recvbuf, op)
+    xc.barrier(context)
+
+    cupy.testing.assert_array_equal(recvbuf, cupy.array(sendbuf * 2))
+
+
+def worker_scatter_buffer(rank, nccl_id, device_id):
+    cupy.cuda.Device(device_id).use()
+
+    context = xc.Context(n_devices=2, nccl_unique_id=nccl_id, rank=rank)
+
+    with pytest.raises(
+        ValueError,
+        match="Buffer size is not exactly divisible by the number of devices",
+    ):
+        not_divisible_buffer = cupy.array([1, 2, 3], dtype=cupy.float32)
+        xc.scatter_buffer(context, not_divisible_buffer)
+
+    with pytest.raises(
+        ValueError,
+        match="Buffer size is not exactly divisible by the number of devices",
+    ):
+        not_divisible_buffer = cupy.array([[1, 2, 3, 4]], dtype=cupy.float32)
+        xc.scatter_buffer(context, not_divisible_buffer)
+
+    buffer = cupy.array([1, 2, 3, 4], dtype=cupy.float32)
+    buffer_chunks = xc.scatter_buffer(context, buffer)
+    assert len(buffer_chunks) == 2
+    cupy.testing.assert_array_equal(buffer_chunks[0], cupy.array([1, 2]))
+    cupy.testing.assert_array_equal(buffer_chunks[1], cupy.array([3, 4]))
+
+    buffer = cupy.array([[1, 2], [3, 4]], dtype=cupy.float32)
+    buffer_chunks = xc.scatter_buffer(context, buffer)
+    assert len(buffer_chunks) == 2
+    cupy.testing.assert_array_equal(buffer_chunks[0], cupy.array([[1, 2]]))
+    cupy.testing.assert_array_equal(buffer_chunks[1], cupy.array([[3, 4]]))
+
+    buffer = cupy.array([[1], [2], [3], [4]], dtype=cupy.float32)
+    buffer_chunks = xc.scatter_buffer(context, buffer)
+    assert len(buffer_chunks) == 2
+    cupy.testing.assert_array_equal(buffer_chunks[0], cupy.array([[1], [2]]))
+    cupy.testing.assert_array_equal(buffer_chunks[1], cupy.array([[3], [4]]))
+
+
+@pytest.mark.parametrize(
+    "worker_func",
+    [
+        worker_gather,
+        worker_all_to_all,
+        worker_allgather,
+        worker_allreduce,
+        worker_broadcast,
+        worker_reduce,
+        worker_reduce_scatter,
+        worker_scatter,
+        worker_barrier,
+        worker_scatter_buffer,
+    ],
+)
+def test_driver(worker_func):
+    mp.set_start_method("spawn", force=True)
+
+    nccl_id = nccl.get_unique_id()
+    device_id = [0, 1]
+
+    process1 = mp.Process(target=worker_func, args=(0, nccl_id, device_id[0]))
+    process1.start()
+    process2 = mp.Process(target=worker_func, args=(1, nccl_id, device_id[1]))
+    process2.start()
+
+    process1.join()
+    process2.join()
+
+    assert process1.exitcode == 0, "Process 1 encountered an error."
```

Comment:
```
>       assert process1.exitcode == 0, "Process 1 encountered an error."
E       AssertionError: Process 1 encountered an error.
E       assert -7 == 0
E        +  where -7 = <Process name='Process-1' pid=118721 parent=118669 stopped exitcode=-SIGBUS>.exitcode

test_cupy.py:254: AssertionError
```
when I debug it， I found that xc.Context.communicator might fail to initialize

Reply:
hmmm i cannot say for sure what the cause is. i have to open separate containers to run the code because i don't have two GPU instances. I also tried to run the tests in a kaggle env since it offers two free T4 GPU, and the tests work as well. but I  never tried to (and was not able to) directly execute the pytest command so it is a bit hard to reproduce. 

regarding SIGBUG, this would be caused by some invalid memory usage or something, and i found some similar issues on nccl as well https://github.com/tloen/alpaca-lora/issues/218 https://github.com/ultralytics/ultralytics/issues/2588

some mentioned solutions were to try using only one type of GPU (I personally only tried T4) or reducing memory usage. I think the later is not needed since the test codes don't require too much memory. but would the error be related to any mixed GPU usage or something?

Reply:
and i guess a bit more context: my cupy version is 12.1.0 

and nvcc:
nvcc: NVIDIA (R) Cuda compiler driver
Copyright (c) 2005-2022 NVIDIA Corporation
Built on Wed_Sep_21_10:33:58_PDT_2022
Cuda compilation tools, release 11.8, V11.8.89
Build cuda_11.8.r11.8/compiler.31833905_0

Reply:
> hmmm i cannot say for sure what the cause is. i have to open separate containers to run the code because i don't have two GPU instances. I also tried to run the code in a kaggle env since it offers two free T4 GPU, and the tests work as well. but I never tried to (and was not able to) directly execute the pytest command so i don't know if things would behave differently under pytest or not.
> 
> If not, SIGBUS could be caused by some invalid GPU memory usage or something, there are some issues regarding SIGBUS on GPU communications:
> 
> [tloen/alpaca-lora#218](https://github.com/tloen/alpaca-lora/issues/218) [ultralytics/ultralytics#2588](https://github.com/ultralytics/ultralytics/issues/2588) [huggingface/diffusers#4229](https://github.com/huggingface/diffusers/issues/4229)
> 
> some mentioned solutions were to try using only one type of GPU (I personally only tried T4) or reducing memory usage. I think the later is not needed since the test code doesn't require a lot of memory. but would the error be related to any mixed GPU usage or something? or people also mention about cuda version etc, i don't know

You are right! I use docker to run it but I don't allocate enough memory for my docker so I account this problem.

Reply:
glad it worked 👍 but yeah feel free to change the code to fit xoscar's need